src/math/exp_fcn.c

```diff
@@ -0,0 +1,171 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2022 Intel Corporation. All rights reserved.
+//
+// Author: Shriram Shastry <malladi.sastry@linux.intel.com>
+//
+//
+
+/* Include Files */
+#include <sof/math/exp_fcn.h>
+#include <sof/math/numbers.h>
+#include <stdint.h>
+#include <stdbool.h>
+
+/* Function Declarations */
+static inline int64_t lomul_s64_SR_sat_near(int64_t a, int64_t b);
+
+static inline void wide_mul_s64(int64_t in_0, int64_t in_1, uint64_t *ptroutbitshi,
+				uint64_t *ptroutbitslo);
+/*
+ * Arguments	: int64_t a
+ *		  int64_t b
+ * Return Type	: int64_t
+ */
+static inline int64_t lomul_s64_SR_sat_near(int64_t a, int64_t b)
+{
+	uint64_t u64_rhi;
+	uint64_t u64_rlo;
+	bool roundup;
+
+	wide_mul_s64(a, b, &u64_rhi, &u64_rlo);
+	roundup = ((u64_rlo & BIT_MASK_LOW_Q27P5) != 0);
+	u64_rlo =
+	    (u64_rhi << 36 | u64_rlo >> 28) + (roundup ? (uint64_t)1 : (uint64_t)0);
+	return (int64_t)u64_rlo;
+}
+
+/*
+ * Arguments	: int64_t in_0
+ *		  int64_t in_1
+ *		  uint64_t *ptroutbitshi
+ *		  uint64_t *ptroutbitslo
+ * Multiplication of two signed int64 numbers
+ * Return Type	: void
+ */
+static inline void wide_mul_s64(int64_t in_0, int64_t in_1, uint64_t *ptroutbitshi,
+				uint64_t *ptroutbitslo)
+{
+	uint64_t absin0;
+	uint64_t absin1;
+	uint64_t in0hi;
+	uint64_t in0lo;
+	uint64_t in1hi;
+	uint64_t prodhilo;
+	uint64_t prodlohi;
+	// bounds
+	absin0 = ABS(in_0);
+	absin1 = ABS(in_1);
+	// High and low bytes
+	in0hi = absin0 >> 32;
+	in0lo = absin0 & UINT32_MAX;
+	in1hi = absin1 >> 32;
+	absin0 = absin1 & UINT32_MAX;
+	// multiplication
+	prodhilo = in0hi * absin0;
+	prodlohi = in0lo * in1hi;
+	absin0 *= in0lo;
+	absin1 = 0;
+	in0lo = absin0 + (prodlohi << 32);
+	if (in0lo < absin0)
+		absin1 = 1;
+
+	absin0 = in0lo;
+	//shift and add
+	in0lo += prodhilo << 32;
+	if (in0lo < absin0)
+		absin1++;
+
+	absin0 = absin1 + in0hi * in1hi + (prodlohi >> 32) + (prodhilo >> 32);
+	if (in_0 != 0 && in_1 != 0 && (in_0 > 0) != (in_1 > 0)) {
+		absin0 = ~absin0;
+		in0lo = ~in0lo;
+		in0lo++;
+		// counter
+		if (in0lo == 0)
+			absin0++;
+	}
+	// pointer- output high and low bytes
+	*ptroutbitshi = absin0;
+	*ptroutbitslo = in0lo;
+}
+
+/* f(x) = a^x, x is variable and a is base
+ *
+ * Arguments	: int32_t x(Q4.28)
+ * input range -5 to 5
+ *
+ * Return Type	: int32_t ts(Q9.23)
+ * output range 0.0067465305 to 148.4131488800
+ *+------------------+-----------------+--------+--------+
+ *| x		     | ts (returntype) |   x    |  ts    |
+ *+----+-----+-------+----+----+-------+--------+--------+
+ *|WLen| FLen|Signbit|WLen|FLen|Signbit| Qformat| Qformat|
+ *+----+-----+-------+----+----+-------+--------+--------+
+ *| 32 | 28  |  1    | 32 | 23 |   0   | 4.28   | 9.23   |
+ *+------------------+-----------------+--------+--------+
+ */
+int32_t exp_int32(int32_t x)
+{
+	/* LUT = ceil(1/factorial(n) * 2 ^ 63) */
+	static const int64_t iv[19] = {
```

Comment:
any xtensa experts know if there is any effects by declaring this in the function?

Reply:
What kind of benefit do you think switching to a function will bring?

Reply:
@marc-hb is correct. @andyross Do you know? If there is no difference might make it a bit cleaner to move it out of the function.

Reply:
Coming late to the call: a static symbol defined internal to a function is treated identically by the code generator and linker to one defined at the top level (the only difference is a prefix that gets added to the name to disambiguate it, since the C symbol is in a different scope).

That said: I personally don't see anything wrong with keeping function local data (even function-local const lookup tables) inside the function.  We don't blink if someone writes `const char *msg = "whatever";` right?  This is exactly the same thing.